### PR TITLE
(FACT-2878) AIX networking resolver rewrite

### PR DESCRIPTION
**Pull request by sebastian-miclea**
Updated AIX resolver to read interfaces using FFI. 
Vlans and secondary ips were not displayed without FFI.

### CODE REVIEW

File: lib/facter/resolvers/aix/ffi/ffi_helper.rb [hunk 1]
```diff
@@ -1,34 +1,152 @@
...
+          averages.read_array_of_long_long(3).map { |x| (x / 65_536.0).round(5) }
+        end
+
+        def self.read_interfaces
+          ksize = Libc.getkerninfo(KINFO_RT_IFLIST, nil, nil, 0)
+
+          ksize_ptr = ::FFI::MemoryPointer.new(:int, ksize.size)
+          ksize_ptr.write_int(ksize)
+
+          count = ksize / 1
```
Comment (GabrielNagy) [inline lib/facter/resolvers/aix/ffi/ffi_helper.rb:45]: does the `/ 1` part change anything in this line?

File: lib/facter/resolvers/aix/ffi/ffi_helper.rb [hunk 2]
```diff
@@ -1,34 +1,152 @@
...
+          result_ptr = ::FFI::MemoryPointer.new(:char, count)
+          res = Libc.getkerninfo(KINFO_RT_IFLIST, result_ptr, ksize_ptr, 0)
+
+          if res == -1
+          end
```
Comment (GabrielNagy) [inline lib/facter/resolvers/aix/ffi/ffi_helper.rb:50]: this might be left over?

File: lib/facter/resolvers/aix/ffi/ffi_helper.rb [hunk 3]
```diff
@@ -1,34 +1,152 @@
...
+
+            case hdr[:ifm_type]
+            when FFI::RTM_IFINFO
+              link_addr = FFI::SockaddrDl.new(hdr.to_ptr + hdr.size)
+
+              interface_name = link_addr[:sdl_data].to_s[0, link_addr[:sdl_nlen]]
+              interfaces[interface_name] ||= {}
+
+            when FFI::RTM_NEWADDR
+              @addresses = {}
```
Comment (GabrielNagy) [inline lib/facter/resolvers/aix/ffi/ffi_helper.rb:65]: does this have to be an instance variable?

**Review by GabrielNagy — COMMENTED** [review]
Just some code-related comments since I have no idea what the code does 😄.
Might be good to have some logging/comments in `ffi_helper.rb`, especially in the `self.read_interfaces` method which has a lot of branching logic.

**Review by ciprianbadescu — APPROVED** [review]
It looks to be the same logic as in facter 3

File: lib/facter/resolvers/aix/ffi/ffi_helper.rb (hunk 1 above)
Comment (sebastian-miclea) [reply lib/facter/resolvers/aix/ffi/ffi_helper.rb:45]: No, I just forgot to remove it.

File: lib/facter/resolvers/aix/ffi/ffi_helper.rb (hunk 2 above)
Comment (sebastian-miclea) [reply lib/facter/resolvers/aix/ffi/ffi_helper.rb:50]: Yes, forgot to remove it.

File: lib/facter/resolvers/aix/ffi/ffi_helper.rb (hunk 3 above)
Comment (sebastian-miclea) [reply lib/facter/resolvers/aix/ffi/ffi_helper.rb:65]: No, I'll update.

**Review by GabrielNagy — APPROVED** [review]

**Pull request merged**